[CI] DissectParserTests.testBasicMatchUnicode

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+intake/240/console

**This reproduces:**
```
./gradlew :libs:dissect:unitTest \
  -Dtests.seed=EDD0A9CCDF94CE9 \
  -Dtests.class=org.elasticsearch.dissect.DissectParserTests \
  -Dtests.method="testBasicMatchUnicode" \
  -Dtests.security.manager=true \
  -Dtests.locale=be \
  -Dtests.timezone=Asia/Karachi \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

```
java.lang.AssertionError: 
Expected: <[10Cc63zu0ymRmfY9Zk1, 3944OrI4JmcvtSZrSXxmMhVMqZDgNWvDezKE5NSI3Dx0BhVGeVMaSr1NF9zLtmlZiBujpCc3vgBhJiHrhddm2bbzb38O, 3z3ajPC8wqxakYDqCqq3tg894G5gKz8O, 4Cb7ehkVzJ2TPgsE0rgdZJU6PB1T, 4UDTbgxjn, BKy3sOon08jqSa3A9GvJJ, CH5an2fWSLfTG4OmAYl4z579WnzjjRViDtyK6TKemQvdXhXECCyLtHga1bFZviZJODPG5AiDEEviba9K, MhpF1MwHnZR0ImHGJ89FY9oGhHEd9uGCrEplvKWO0fidqGqqaX9HTJhh0czDWzYB89gOOgkjebOZ8pfdI84f7wLLbwH0kY8RO, RJ4K, Vgnp4mfwnLoGnD3UzOli27qb3y8JAVkY7CVWreu5r1RXE5KrSsP9DADb5pU478uTnv17Wg4PtaU1tbofEU, a08Be14bgaXvExDtLhduVn3Edo77pVtGtSOfe4KkxdDQovFYVBtEJjy0YXkd4Xwl78hfyRQbHSxr1mMSvwb9qKbyMsRo4, e, e, gCFLt2bweKr9zYkpqlwGRvnyjavnBZcI9YjxnhOE2kCWzRHnnqBQDu0yN2PFQHYyVaShvTNcD1T3o1NM3IKfUZpsng1kde, ghzUz, kdJ42wwuzr9Uqq5anJKKEq2rFTLziIBUjrfkjnUjaowFmE786SlY15fq15wr8kWsuA8AsbGL2vddTdDSKYksEvJeyAulUIN3, wMpYABhJ5VvTG2WOzP6gpIbw4DgagJtCY45imUGiOzIa1TLByAHd8B4jnKY0NEsoiJVYrrAkLYBB7BqgvqHlRQ37xMGu8bwOK8iZ, yVybK2, yboKhUQUsj1xaZQMCkRt1qr5PJyVjolC8eRVEUNsr2ud3CfpThmKQXPT]>
     but: was <[10Cc63zu0ymRmfY9Zk1, 3944OrI4JmcvtSZrSXxmMhVMqZDgNWvDezKE5NSI3Dx0BhVGeVMaSr1NF9zLtmlZiBujpCc3vgBhJiHrhddm2bbzb38O, 3z3ajPC8wqxakYDqCqq3tg894G5gKz8O, 4Cb7ehkVzJ2TPgsE0rgdZJU6PB1T, 4UDTbgxjn, BKy3sOon08jqSa3A9GvJJ, CH5an2fWSLfTG4OmAYl4z579WnzjjRViDtyK6TKemQvdXhXECCyLtHga1bFZviZJODPG5AiDEEviba9K, MhpF1MwHnZR0ImHGJ89FY9oGhHEd9uGCrEplvKWO0fidqGqqaX9HTJhh0czDWzYB89gOOgkjebOZ8pfdI84f7wLLbwH0kY8RO, RJ4K, Vgnp4mfwnLoGnD3UzOli27qb3y8JAVkY7CVWreu5r1RXE5KrSsP9DADb5pU478uTnv17Wg4PtaU1tbofEU, a08Be14bgaXvExDtLhduVn3Edo77pVtGtSOfe4KkxdDQovFYVBtEJjy0YXkd4Xwl78hfyRQbHSxr1mMSvwb9qKbyMsRo4, e, gCFLt2bweKr9zYkpqlwGRvnyjavnBZcI9YjxnhOE2kCWzRHnnqBQDu0yN2PFQHYyVaShvTNcD1T3o1NM3IKfUZpsng1kde, ghzUz, kdJ42wwuzr9Uqq5anJKKEq2rFTLziIBUjrfkjnUjaowFmE786SlY15fq15wr8kWsuA8AsbGL2vddTdDSKYksEvJeyAulUIN3, wMpYABhJ5VvTG2WOzP6gpIbw4DgagJtCY45imUGiOzIa1TLByAHd8B4jnKY0NEsoiJVYrrAkLYBB7BqgvqHlRQ37xMGu8bwOK8iZ, yVybK2, yboKhUQUsj1xaZQMCkRt1qr5PJyVjolC8eRVEUNsr2ud3CfpThmKQXPT]>
	at __randomizedtesting.SeedInfo.seed([EDD0A9CCDF94CE9:10BEBC69DFFCF02B]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.dissect.DissectParserTests.assertMatch(DissectParserTests.java:383)
	at org.elasticsearch.dissect.DissectParserTests.assertMatch(DissectParserTests.java:372)
	at org.elasticsearch.dissect.DissectParserTests.testBasicMatchUnicode(DissectParserTests.java:149)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.lang.Thread.run(Thread.java:748)
```
I'll be subtly snatching this from the ~ @elastic/es-search team~ @elastic/es-core-features  , and try to fix it!

--- a/libs/dissect/src/test/java/org/elasticsearch/dissect/DissectParserTests.java
+++ b/libs/dissect/src/test/java/org/elasticsearch/dissect/DissectParserTests.java
@@ -28,7 +28,6 @@ import org.mockito.internal.util.collections.Sets;
 
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;

--- a/libs/dissect/src/test/java/org/elasticsearch/dissect/DissectParserTests.java
+++ b/libs/dissect/src/test/java/org/elasticsearch/dissect/DissectParserTests.java
@@ -24,6 +24,7 @@ import com.fasterxml.jackson.databind.ObjectMapper;
 import org.elasticsearch.test.ESTestCase;
 import org.hamcrest.CoreMatchers;
 import org.hamcrest.Matchers;
+import org.mockito.internal.util.collections.Sets;
 
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -112,7 +113,7 @@ public class DissectParserTests extends ESTestCase {
         String delimiterFirstInput = "";
         String delimiterFirstPattern = "";
         //parallel arrays
-        List<String> expectedKeys = Arrays.asList(generateRandomStringArray(100, 10, false, false));
+        List<String> expectedKeys = new ArrayList<>(Sets.newSet(generateRandomStringArray(100, 10, false, false)));
         List<String> expectedValues = new ArrayList<>(expectedKeys.size());
         for (String key : expectedKeys) {
             String value = randomAsciiAlphanumOfLengthBetween(1, 100);
@@ -127,7 +128,6 @@ public class DissectParserTests extends ESTestCase {
         assertMatch(delimiterFirstPattern, delimiterFirstInput, expectedKeys, expectedValues);
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/39244")
     public void testBasicMatchUnicode() {
         String valueFirstInput = "";
         String keyFirstPattern = "";
@@ -138,6 +138,9 @@ public class DissectParserTests extends ESTestCase {
         List<String> expectedValues = new ArrayList<>();
         for (int i = 0; i < randomIntBetween(1, 100); i++) {
             String key = randomAsciiAlphanumOfLengthBetween(1, 100);
+            while (expectedKeys.contains(key)) { // keys should be unique in this test
+                key = randomAsciiAlphanumOfLengthBetween(1, 100);
+            }
             String value = randomRealisticUnicodeOfCodepointLengthBetween(1, 100);
             String delimiter = Integer.toString(randomInt()); //int to ensures values and delimiters don't overlap, else validation can fail
             keyFirstPattern += "%{" + key + "}" + delimiter;
@@ -375,13 +378,11 @@ public class DissectParserTests extends ESTestCase {
 
     private void assertMatch(String pattern, String input, List<String> expectedKeys, List<String> expectedValues, String appendSeperator) {
         Map<String, String> results = new DissectParser(pattern, appendSeperator).parse(input);
-        List<String> foundKeys = new ArrayList<>(results.keySet());
-        List<String> foundValues = new ArrayList<>(results.values());
-        Collections.sort(foundKeys);
-        Collections.sort(foundValues);
-        Collections.sort(expectedKeys);
-        Collections.sort(expectedValues);
-        assertThat(foundKeys, Matchers.equalTo(expectedKeys));
-        assertThat(foundValues, Matchers.equalTo(expectedValues));
+        assertThat(results.size(), Matchers.equalTo(expectedKeys.size()));
+        assertThat(results.size(), Matchers.equalTo(expectedValues.size()));
+        for (int i = 0; i < results.size(); i++) {
+            final String key = expectedKeys.get(i);
+            assertThat(results.get(key), Matchers.equalTo(expectedValues.get(i)));
+        }
     }
 }

